src/bin/pgcopydb/pgsql.c

```diff
@@ -1843,6 +1869,216 @@ pgsql_fetch_results(PGSQL *pgsql, bool *done,
+{
+#if defined(LIBPQ_HAS_PIPELINING) && LIBPQ_HAS_PIPELINING
+	PGconn *conn = pgsql->connection;
+
+	if (conn == NULL)
+	{
+		log_error("BUG: pgsql_sync_pipeline called with NULL connection");
+		return false;
+	}
+
+	log_trace("Start pipeline sync");
+
+	if (PQpipelineStatus(conn) != PQ_PIPELINE_ON)
+	{
+		log_error("BUG: Connection is not in pipeline mode");
+		return false;
+	}
+
+	if (PQsocket(conn) < 0)
+	{
+		(void) pgsql_stream_log_error(pgsql, NULL, "invalid socket");
+
+		clear_results(pgsql);
+		pgsql_finish(pgsql);
+
+		return false;
+	}
+
+	if (PQpipelineSync(conn) != 1)
+	{
+		(void) pgcopy_log_error(pgsql, NULL, "Failed send sync pipeline");
+		return false;
+	}
+
+	bool syncReceived = false;
+
+	while (!syncReceived)
+	{
+		if (asked_to_quit || asked_to_stop || asked_to_stop_fast)
+		{
+			log_error("Pipeline sync was interrupted");
+
+			clear_results(pgsql);
+			pgsql_finish(pgsql);
+
+			return false;
+		}
+
+		fd_set input_mask;
+		struct timeval timeout;
+		struct timeval *timeoutptr = NULL;
+
+		/* sleep for 10ms to wait for input on the Postgres socket */
+		timeout.tv_sec = 0;
+		timeout.tv_usec = 10000;
+		timeoutptr = &timeout;
+
+		FD_ZERO(&input_mask);
+		FD_SET(PQsocket(conn), &input_mask);
+
+		int r = select(PQsocket(conn) + 1, &input_mask, NULL, NULL, timeoutptr);
+
+		if (r == 0 || (r < 0 && errno == EINTR))
+		{
+			/* got a timeout or signal. The caller will get back later. */
+			continue;
+		}
+		else if (r < 0)
+		{
+			(void) pgsql_stream_log_error(pgsql, NULL, "select failed: %m");
+
+			clear_results(pgsql);
+			pgsql_finish(pgsql);
+
+			return false;
+		}
+
+		/* Else there is actually data on the socket */
+		if (PQconsumeInput(conn) == 0)
+		{
+			(void) pgsql_stream_log_error(
+				pgsql,
+				NULL,
+				"Failed to consume input for pipeline sync");
+			return false;
+		}
+
+		int results = 0;
+
+		while (!PQisBusy(conn))
+		{
+			PGresult *res = PQgetResult(conn);
+
+			if (res == NULL)
+			{
+				/*
+				 * NULL represents a end of result for a single query, but we are
+				 * in pipeline mode and we can have multiple results.
+				 *
+				 * We need to continue consuming until we get a SYNC message.
+				 */
+
+				continue;
+			}
+
+			results++;
+
+			ExecStatusType resultStatus = PQresultStatus(res);
+
+			PQclear(res);
+
+			if (resultStatus == PGRES_PIPELINE_SYNC)
+			{
+				syncReceived = true;
+
+				log_trace("Received pipeline sync. Total results: %d", results);
+
+				break;
+			}
+
+			if (!is_response_ok(res))
+			{
+				(void) pgcopy_log_error(pgsql, res, "Failed to receive pipeline sync");
+				return false;
+			}
+		}
+	}
+
+	/* update the last pipeline sync time */
+	pgsql->pipelineSyncTime = time(NULL);
+
+	log_trace("Endof pipeline sync");
+
+	return true;
+#else
+	static bool warned = false;
+
+	if (!warned)
+	{
+		log_warn("libpq in Postgres version %d does not support pipeline mode, "
+				 "need at least version 14",
+				 PG_MAJORVERSION_NUM);
+		warned = true;
```

Comment:
I still think we should WARN at build-time using the `#warning` directive, and rewrite the error-message in a way that makes sense for our run-time users:

``` c
#warning "Compiling pgcopydb without pipeline mode support"
    log_warn("Skipping libpq pipeline mode optimisation because pgcopydb was built with libpq %d, "
             "pipeline mode is available since libpq 14");
```

Please also see if we can add libpq version in the `#warning` message?

Reply:
@dimitri I've already warning directive here => https://github.com/dimitri/pgcopydb/pull/704/files#diff-84ab9d926dfb0eda4b691a03c25ba22211e96c9f749fa57ff44ca24758bb5777R36. 

I will change the message as per your suggestion. thanks!